### PR TITLE
AST/HIR: Introduce `ExprKind::Err` for better error recovery in the front-end

### PR DESCRIPTION
**Pull request by petrochenkov**
This way we can avoid aborting compilation if expansion produces errors and generate `ExprKind::Err`s instead.

### CODE REVIEW

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -1284,16 +1284,15 @@ impl<'a> Context<'a> {
...
-            if !is_macro {
+        if !attr::is_known(attr) {
+            if name.starts_with("rustc_") {
+                let msg = "unless otherwise specified, attributes with the prefix `rustc_` \
+                           are reserved for internal compiler diagnostics";
+                gate_feature!(self, rustc_attrs, attr.span, msg);
+            } else if !is_macro {
+                // Only run the custom attribute lint during regular feature gate
+                // checking. Macro gating runs before the plugin attributes are
+                // registered, so we skip this in that case.
```
Comment (estebank) [inline src/libsyntax/feature_gate.rs]: Was this change _needed_? I don't mind it, just curious if it changed anything.

**Review by estebank — COMMENTED** [review]
Really cool!

CC #48724

File: src/test/ui/did_you_mean/issue-40396.stderr [hunk 2]
```diff
@@ -1,38 +1,107 @@
...
    |
    = help: use `::<...>` instead of `<...>` if you meant to specify type arguments
    = help: or use `(...)` if you meant to specify fn arguments
 
-error: aborting due to 4 previous errors
+error[E0423]: expected value, found struct `Vec`
+  --> $DIR/issue-40396.rs:12:21
+   |
+LL |     (0..13).collect<Vec<i32>>();
+   |                     ^^^ did you mean `Vec { /* fields */ }`?
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/test/ui/did_you_mean/issue-40396.stderr]: This is pretty bad, but not actually a fallout from this PR.
`(0..13).collect<Vec<i32>>()` outside of macros already produced this.
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0be590d180bad8cbea3789d5e7d2e269

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (petrochenkov) [reply src/libsyntax/feature_gate.rs]: Yes, it's to get rid of duplicate ``attributes with the prefix `rustc_` are reserved`` errors.

Comment (petrochenkov) [reply src/libsyntax/feature_gate.rs]: Similarly to few other changes in https://github.com/rust-lang/rust/pull/56999/commits/21370386f370c828fe5378387ee7223a931e1480.
The diff looked pretty bad when I *just* removed the `abort_if_error`, all the other stuff is attempts to reduce noise from duplicated or nonsensical errors.

File: src/libsyntax/print/pprust.rs [hunk 3]
```diff
@@ -1102,7 +1102,9 @@ impl<'a> State<'a> {
                 self.s.word("_")?;
             }
             ast::TyKind::Err => {
-                self.s.word("?")?;
+                self.popen()?;
+                self.s.word("/*ERROR*/")?;
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/libsyntax/print/pprust.rs]: You're absolutely right.

**Comment by petrochenkov** [comment]
Updated with test fixes and one new commit making sure that all feature-gate errors are recoverable.

**Comment by estebank** [comment]
r=me once travis is happy

**Comment by Centril** [comment]
This PR failed in a rollup: https://github.com/rust-lang/rust/pull/57075#issuecomment-449636661

**Pull request merged**